File: src/tests/file-c/lseek.c
#include <assert.h>
#include <fcntl.h>
#include <limits.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

//==================================================================================================
// Constants
//==================================================================================================

// Maximum length for file content.
#define DATA_LEN_MAX 256

//==================================================================================================
// Standalone Functions
//==================================================================================================

// Tests whether we can manipulate file offsets using lseek.
void test_lseek(void)
{
    fprintf(stderr, "testing lseek() ... ");

    const char *filename = "testfile.tmp";
    assert(strlen(filename) <= NAME_MAX);

    // Data to write using write.
    const char *data1 = "Hello ";
    const char *data2 = "Nanvix!";
    size_t data1_len = strlen(data1);
    size_t data2_len = strlen(data2);
    assert(data1_len + data2_len <= DATA_LEN_MAX);

    char buffer[DATA_LEN_MAX + 1];

    // Create and open a test file.
    int fd = open(filename, O_CREAT | O_RDWR, S_IRUSR | S_IWUSR);
    assert(fd != -1);

    // Write data1 to the file.
    ssize_t bytes_written = write(fd, data1, data1_len);
    assert(bytes_written == (ssize_t)data1_len);

    // Use lseek to move the file offset to the end of data1.
    off_t offset = lseek(fd, 0, SEEK_END);
    assert(offset == (off_t)data1_len);

    // Write data2 at the current offset.
    bytes_written = write(fd, data2, data2_len);
    assert(bytes_written == (ssize_t)data2_len);

    // Use lseek to move the file offset back to the beginning.
    offset = lseek(fd, 0, SEEK_SET);
    assert(offset == 0);

    // Read the data back.
    ssize_t bytes_read = read(fd, buffer, data1_len + data2_len);
    assert(bytes_read == (ssize_t)(data1_len + data2_len));

    // Null-terminate the buffer and assert contents.
    buffer[bytes_read] = '\0';
    assert(strcmp(buffer, "Hello Nanvix!") == 0);

    // Close the file.
    assert(close(fd) == 0);

    // Remove the test file.
    assert(unlink(filename) == 0);

    fprintf(stderr, "passed\n");
}

File: src/tests/integration/test-rust-file/src/unsafe/file/lseek.rs
// Copyright(c) The Maintainers of Nanvix.
// Licensed under the MIT License.

//==================================================================================================
// Imports
//==================================================================================================

use ::sysapi::ffi::c_int;
use ::syscall::{
    fcntl,
    unistd,
};
use sysapi::{
    sys_stat::file_mode::{
        S_IRUSR,
        S_IWUSR,
    },
    sys_types::{
        c_size_t,
        mode_t,
        off_t,
    },
    unistd::file_seek::{
        SEEK_END,
        SEEK_SET,
    },
};
use syscall::safe::RegularFileOpenFlags;

//==================================================================================================
// Standalone Functions
//==================================================================================================

/// Tests whether we can manipulate the seek position of a file.
pub fn test() {
    const DATA: &[u8] = b"Hello Nanvix!";
    let filename: &str = "test-lseek.txt";

    // Create a file and assert result.
    let mode: mode_t = S_IRUSR | S_IWUSR;
    let fd: c_int = match fcntl::creat(filename, mode) {
        Ok(fd) => fd,
        Err(error) => {
            panic!("{error:?}");
        },
    };

    // Close file and assert result.
    if let Err(error) = unistd::close(fd) {
        panic!("{error:?}");
    }

    // Open file for reading and writing and assert result.
    let fd: c_int = match fcntl::open(filename, RegularFileOpenFlags::read_write().into(), 0) {
        Ok(fd) => fd,
        Err(error) => {
            panic!("{error:?}");
        },
    };

    // Write tot he beginning of the file and assert result.
    match unistd::write(fd, DATA) {
        Ok(n) if usize::try_from(n).ok() == Some(DATA.len()) => {},
        Ok(n) => {
            panic!("expected to write {} bytes, but wrote {n} bytes", DATA.len());
        },
        Err(error) => {
            panic!("{error:?}");
        },
    }

    // Seek to the beginning of the file and assert result.
    if let Err(error) = unistd::lseek(fd, 0, SEEK_SET) {
        panic!("{error:?}");
    }

    // Read from the file and assert result.
    let mut expected_data: [u8; DATA.len()] = [0; DATA.len()];
    match unistd::read(fd, &mut expected_data) {
        Ok(n) if usize::try_from(n).ok() == Some(DATA.len()) => {
            if expected_data != DATA {
                panic!("expected to read {:?}, but read {:?}", DATA, expected_data);
            }
        },
        Ok(n) => {
            panic!("expected to read {} bytes, but read {n} bytes", DATA.len());
        },
        Err(error) => {
            panic!("{error:?}");
        },
    }

    // Close file and assert result.
    if let Err(error) = unistd::close(fd) {
        panic!("{error:?}");
    }

    // Open file for reading and writing again and assert result.
    let fd: c_int = match fcntl::open(filename, RegularFileOpenFlags::read_write().into(), 0) {
        Ok(fd) => fd,
        Err(error) => {
            panic!("{error:?}");
        },
    };

    // Seek to the end of the file and assert result.
    if let Err(error) = unistd::lseek(fd, 0, SEEK_END) {
        panic!("{error:?}");
    }

    // Write to the end of the file and assert result.
    match unistd::write(fd, DATA) {
        Ok(n) if usize::try_from(n).ok() == Some(DATA.len()) => {},
        Ok(n) => {
            panic!("expected to write {} bytes, but wrote {n} bytes", DATA.len());
        },
        Err(error) => {
            panic!("{error:?}");
        },
    }

    // Rewind from the end of the file.
    let offset: i64 = match off_t::try_from(DATA.len()) {
        Ok(offset) => -offset,
        Err(_) => {
            panic!("failed to convert length to i32");
        },
    };
    if let Err(error) = unistd::lseek(fd, offset, SEEK_END) {
        panic!("{error:?}");
    }

    // Read data back and assert result.
    let expected_size: c_size_t = match DATA.len().try_into() {
        Ok(size) => size,
        Err(_) => {
            panic!("failed to convert data length to size_t");
        },
    };
    let mut expected_data: [u8; DATA.len()] = [0; DATA.len()];
    match unistd::read(fd, &mut expected_data) {
        Ok(n) if n == expected_size => {
            assert_eq!(&expected_data[..DATA.len()], DATA);
        },
        Ok(n) => {
            panic!("expected to read {} bytes, but read {n} bytes", DATA.len());
        },
        Err(error) => {
            panic!("{error:?}");
        },
    }

    // Close file and assert result.
    if let Err(error) = unistd::close(fd) {
        panic!("{error:?}");
    }

    // Unlink the file and assert result.
    if let Err(error) = unistd::unlink(filename) {
        panic!("{error:?}");
    }
}
